extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/context/ArcContextProvider.java

```diff
@@ -34,59 +39,18 @@ public ThreadContextSnapshot currentContext(Map<String, String> map) {
-                if (arcContainer == null || !arcContainer.isRunning()) {
-                    //this happens on shutdown, if we blow up here it can break shutdown, and stop
-                    //resources from being cleaned up, causing tests to fail
-                    return NOOP_CONTROLLER;
-                }
-                ThreadContextController controller;
-                ManagedContext requestContext = arcContainer.requestContext();
-                // this is executed on another thread, context can but doesn't need to be active here
-                if (ArcContextProvider.this.isContextActiveOnThisThread(arcContainer)) {
-                    // context active, store current state, feed it new one and restore state afterwards
-                    InjectableContext.ContextState stateToRestore = requestContext.getState();
-                    requestContext.deactivate();
-                    if (state != null) {
-                        // only activate if previous thread had it active
-                        requestContext.activate(state);
-                    }
-                    controller = new ThreadContextController() {
-                        @Override
-                        public void endContext() throws IllegalStateException {
-                            // clean up, reactivate context with previous values
-                            if (state != null) {
-                                // only deactivate if previous thread had it active
-                                requestContext.deactivate();
-                            }
-                            requestContext.activate(stateToRestore);
-                        }
-                    };
-                } else {
-                    // context not active, activate and pass it new instance, deactivate afterwards
-                    if (state != null) {
-                        // only activate if previous thread had it active
-                        requestContext.activate(state);
-
-                    }
-                    controller = new ThreadContextController() {
-                        @Override
-                        public void endContext() throws IllegalStateException {
-                            if (state != null) {
-                                // only deactivate if previous thread had it active
-                                requestContext.deactivate();
-                            }
-                        }
-                    };
-                }
-                return controller;
-            }
-        };
+        InjectableContext.ContextState state = null;
+        ManagedContext requestContext = arc.requestContext();
+        if (requestContext.isActive()) {
+            state = requestContext.getState();
+        }
+        if (state == null) {
+            return NULL_CONTEXT_SNAPSHOT;
+        } else if (state.isEmpty()) {
+            return EMPTY_CONTEXT_SNAPSHOT;
```

Comment:
I think this is actually wrong - you won't share the `ContextState` which mean any changes into the context on new thread (newly created beans) won't propagate back to the original active context. E.g. the "backward propagation" will break. 

But scrambling through the code, I don't this this is tested on MP CP level(which it should be), there is only ArC `ContextState` test [here](https://github.com/quarkusio/quarkus/blob/master/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/contexts/request/propagation/RequestContextPropagationTest.java).

Reply:
Yes, you're right and it's a pity... I'll update the PR accordingly :-(